### ISSUE TITLE
[🚀 Feature]: Implement explicit flag in Selenium Manager for managing drivers

### ISSUE THREAD

**Issue by bonigarcia** [SeleniumHQ/selenium#11684]
### Feature and motivation

So far, the main feature of Selenium Manager is the management (download, cache) of drivers (chromedriver, geckodriver, etc.). This was done explicitly, simply specifying a browser (or driver) name, for instance:

```
selenium-manager --browser chrome
selenium-manager --driver chromedriver
```

The next major feature to be implemented in Selenium Manager is the automated management (download, cache) of the principal browsers nowadays: Chrome, Firefox, and Edge.

With this new feature, some Selenium Manager flags (such as `--browser` or `--browser-version`) will be reused for browser management (which is specified using the flag `--resolve-browser`).

When this new flag `--resolve-browser`, maybe it is a good idea to distinguish the driver management with an equivalent flag, such as `--resolve-driver`.

This change would require to implement the flag in the Rust-side, and also change how the bindings call Selenium Manager for managing drivers.

### Usage example

```
selenium-manager --browser chrome --resolve-driver
```

**Comment by titusfortner** [comment]
Can we just have `--browser` and `--browser-version` be for browsers and `--driver` and `--driver-version` be for drivers? We don't have to maintain backwards compatibility (yet).

**Comment by bonigarcia** [comment]
The flag `--browser-version` should be used for driver management as well, IMO. The way I see it, this is one of the key features of Selenium Manager, i.e., given a major browser version (e.g. Chrome 101), Selenium Manager resolves the proper driver version (e.g. chromedriver 101.x.x.x).

**Comment by titusfortner** [comment]
Right. Let me rephrase.

Selenium Bindings are going to take the user's Options class, and send to Selenium Manager: `--browser <browserName> --browser-version <browserVersion>`. That should check to see if the browser with that version exists on the machine & download it if not, and then update the driver to the latest driver supporting that version.

That should be the default behavior without needing additional "resolve" toggles. fwiw, I'm no longer interested in what `--driver` or `--driver-version` do.

**Comment by bonigarcia** [comment]
Yes, we can continue doing driver management in an implicit way, like it is done now. I am proposing this second flag because when Selenium Manager also manages browsers, we might want another flag to distinguish both features. Let me put some examples:

```
# Implicit driver management (it will manage chromedriver)
selenium-manager --browser chrome 
```

```
# Explicit browser management (it will manage Chrome)
selenium-manager --browser chrome --resolve-browser
```

```
# Implicit driver management (it will manage chromedriver) (works exactly the same as implicit driver management)
selenium-manager --browser chrome  --resolve-driver
```

The benefit I see of having two separate flags for browsers and drivers is that you can use both of them in the same call (to manage both the browser and the driver):

```
# Implicit driver and browser management (it will manage Chrome and chromedriver)
selenium-manager --browser chrome  --resolve-driver --resolve-browser
```

Alternatively, you might call Selenium Manager twice (one for the browser and the second for the driver). Or implicitly, resolve the driver any time you resolve the browser.

Maybe the best choice is to have a `--resolve-driver` flag which is implicit for driver management but explicit for driver management when done at the same time of browser management.

**Comment by titusfortner** [comment]
This needs to explicitly manage the browser and implicitly manage the driver:

```
selenium-manager --browser chrome
```

These features need to center the Selenium Bindings as the target user, and that means the default should be what the Bindings need. If someone somewhere at some point needs different features for some reason, then we can provide toggles to support them.

**Comment by diemol** [comment]
I do not think this is needed. Selenium Manager should try to setup the environment and put in place what is missing. Which means download browser and driver, or only one of them when it applies.

What makes more sense to discuss, not sure if in this issue or another one, is the logic Selenium Manager will follow for different use cases, like what happens if I do `selenium-manager --browser chrome` and the user has ChromeDriver 99 on the path, do we error or download latest Chrome and update ChromeDriver? Etc.

**Comment by titusfortner** [comment]
* `selenium-manager --browser chrome` grabs existing chrome if found; otherwise downloads latest stable browser and uses it, and updated driver for it
* `selenium-manager --browser chrome --browser-version <channel>` looks for browser with that channel, if it doesn't find and/or what it finds is out of date, it will download and use along with the updated driver. I think we would support `latest`, `stable`, `beta`, `dev`, `canary` (though the last gets weird because of how the drivers get distributed)

**Comment by bonigarcia** [comment]
> Selenium Manager should try to setup the environment and put in place what is missing. Which means download browser and driver, or only one of them when it applies.

It can be done in that way as well.

Maybe at some point in the future, we want to disable the driver or browser management (using, for instance, `--disable-driver` and `--disable-browser` flags), just in case only one management type is required.

> * `selenium-manager --browser chrome` grabs existing chrome if found; otherwise downloads latest stable browser and uses it, and updated driver for it
> * `selenium-manager --browser chrome --browser-version <channel>` looks for browser with that channel, if it doesn't find and/or what it finds is out of date, it will download and use along with the updated driver. I think we would support `latest`, `stable`, `beta`, `dev`, `canary` (though the last gets weird because of how the drivers get distributed)

Yes, it can be done like that. In that case, we need to decide what is the output of Selenium Manager concerning browser/driver management. Currently, the output of Selenium Manager is the absolute path of the resolved driver. When the browser management is done as well, we need to establish a convention about the Selenium Manager output. Maybe the first line output line (INFO) will be the browser path, and the second line will be the driver path. That will be reflected in the JSON output as well.

**Comment by titusfortner** [comment]
Yeah, that was the main reason I supported the json output, we need it to output the location of both driver and browser. Bindings can parse those two values.

If a browser is downloaded, the bindings will have to update the browser options instance with its location.

I'm not sure the right answer for the other output types.  Shell output for --browser gives browser location and for --driver gives driver location?

**Comment by bonigarcia** [comment]
> I'm not sure the right answer for the other output types. Shell output for --browser gives browser location and for --driver gives driver location?

That is a convention we need to define. I would say that both `--browser` and `--driver` return both the browser and driver location.

**Comment by titusfortner** [comment]
We created the shell output toggle so that we could pipe the result directly somewhere else, so I don't think we want to return both in that mode. The other modes we can.

**Comment by bonigarcia** [comment]
I think this issue is no longer required. Closing.
